### DISCUSSION TITLE
First time I see a library with static typing for the database

### DISCUSSION THREAD

**Discussion by Sleepful** [diesel-rs/diesel#4073] [Ideas]
This is the first time I see a library (in general for any language) that provides some actual strong guarantees in terms of static-typing deserealization from database rows. 

I am just, shocked? This is amazing, why is this the first time I ever see this idea put into practice? Even other statically typed languages rely on correctly and manually hand-typing the types for deserealized DB data, which has been a source of much bugs and pain.

Are there any libraries that do anything remotely similar? A db-first type-generation. The closest thing is something like Prisma and even then, that's still not it. Prisma is a code-schema-first, and _then_ it migrates the database to match the code. 

Diselrs simply blew my mind.

**Comment by weiznich** [comment]
Thanks for the kind words. It's always great to hear that people like the crate. 

I'm aware of the following other libraries that provide a similar approach:  

* [esqueleto (Haskel)](https://hackage.haskell.org/package/esqueleto)
* [sqlx (rust)](https://docs.rs/sqlx/latest/sqlx/) (although that's a slightly different approach again, that has a few practical limitations to what can be inferred)
* [sqlc (go)](https://sqlc.dev/) (similar approach as sqlx with similar limitations)
